python/dialects/aiex.py

```diff
@@ -836,3 +836,64 @@ def dma_start_bd_chain_for(symbol, args, alloc, *pyargs, **kwargs):
     return DMAStartBdChainForOp(
         T.index(), chain_sym, args, alloc_sym, *pyargs, **kwargs
     )
+
+
+def shim_dma_bd(
+    mem,
+    offset: int | None = None,
+    sizes: MixedValues | None = None,
+    strides: MixedValues | None = None,
+    transfer_len: int | None = None,
+):
+    if offset is None:
+        offset = 0
+    if sizes is None:
+        sizes = [0] * 4
+    if strides is None:
+        strides = [0] * 3 + [1]
```

Comment:
`len` is a special function in python, e.g.,
```python
my_list = [1, 2, 3]
len(my_list)
```

It's best not to override this. We should never use `len` for arguments in python bindings, if we can avoid it.

Reply:
If we don't know the sizes, we can't calculate the length for the user. So I believe there should be defaults at the python level, unless we move length calculation lower.

Reply:
> If we don't know the sizes, we can't calculate the length for the user. So I believe there should be defaults at the python level, unless we move length calculation lower.
> 
> This code was also copy/pasted from the `npu_dma_memcpy_nd` python wrapper. I think it's important for users to have consistent behavior between interfaces that do similar things, if possible.

I won't pretend I've ingested all the details here but consistent interfaces are important. One way to get that is to put things at the C++ level so that iron, normal mlir-aie, _as well as other users_ (e.g. mlir-air) all see the same behavior.

Reply:
@andrej @jgmelber are there any technical reasons you can think of why the length cannot be calculated at a lower level if the user does not specify a length?

Reply:
I believe for dma_memcpy_nd we already have the length calculation at the lower level. So no I don't think there's anything preventing us from doing it at the lower level.